pvlib/ivtools.py

```diff
@@ -262,6 +268,173 @@ def fit_sde_sandia(voltage, current, v_oc=None, i_sc=None, v_mp_i_mp=None,
+
+    Tref = temp_ref + 273.15  # in K
+
+    if celltype.lower() in ['monosi', 'polysi', 'multisi']:
+        C = 0.0002677  # valid for silicon
+        Eg = 1.796e-19  # in J, valid for silicon
+    elif celltype.lower() in ['gaas']:
+        C = 0.0003174  # valid for gallium arsenide
+        Eg = 2.163e-19  # in J, valid for gallium arsenide
+    elif celltype.lower() in ['cis', 'cigs', 'cdte', 'amorphous']:
+        raise NotImplementedError
+    else:
+        raise ValueError('Unknown cell type.')
+
+    # Conversion from %/K to A/K & V/K
+    alpha_sc = alpha_sc*i_sc/100
+    beta_voc = beta_voc*v_oc/100
+
+    def pv_fct(params, specs):
+        """Returns the system of equations used for computing the
+        single-diode 5 parameters.
+        To avoid the confusion in names with variables of container
+        function the '_' of the variables were removed.
+        """
+        # six input known variables
+        Isc = specs[0]
+        Voc = specs[1]
+        Imp = specs[2]
+        Vmp = specs[3]
+        betaoc = specs[4]
+        alphasc = specs[5]
+        # five parameters vector to find
+        IL = params[0]
+        Io = params[1]
+        a = params[2]
+        Rsh = params[3]
+        Rs = params[4]
+        # five equation vector
+        y = [0, 0, 0, 0, 0]
+
+        # 1st equation - short-circuit - eq(3) in [1]
+        y[0] = Isc - IL + Io*(np.exp(Isc*Rs/a) - 1.0) + Isc*Rs/Rsh
+        # 2nd equation - open-circuit Tref - eq(4) in [1]
+        y[1] = -IL + Io*(np.exp(Voc/a) - 1.0) + Voc/Rsh
+        # 3rd equation - Imp & Vmp - eq(5) in [1]
+        y[2] = Imp - IL + Io*(np.exp((Vmp+Imp*Rs)/a) - 1.0) + \
+            (Vmp+Imp*Rs)/Rsh
+        # 4th equation - Pmp derivated=0 -
+        # caution: eq(6) in [1] seems to be incorrect, take eq23.2.6 in [2]
+        y[3] = Imp - Vmp * ((Io/a)*np.exp((Vmp+Imp*Rs)/a) + 1.0/Rsh) / \
+            (1.0 + (Io*Rs/a)*np.exp((Vmp+Imp*Rs)/a) + Rs/Rsh)
+        # 5th equation - open-circuit T2 - eq (4) at temperature T2 in [1]
+        T2 = Tref + 2
+        Voc2 = (T2 - Tref)*betaoc + Voc  # eq (7) in [1]
+        a2 = a*T2/Tref  # eq (8) in [1]
+        IL2 = IL + alphasc*(T2-Tref)  # eq (11) in [1]
+        Eg2 = Eg*(1-C*(T2-Tref))  # eq (10) in [1]
+        Io2 = Io * (T2/Tref)**3 * np.exp(1/k * (Eg/Tref-Eg2/T2))  # eq (9)
+        y[4] = -IL2 + Io2*(np.exp(Voc2/a2) - 1.0) + Voc2/Rsh  # eq (4) at T2
+
+        return y
+
+    # initial guesses of variables for computing convergence:
+    # Values are taken from [2], p753
+    Rsh_i = 100.0
+    a_i = 1.5*k*Tref*cells_in_series/q
+    IL_i = i_sc
+    Io_i = i_sc * np.exp(-v_oc/a_i)
+    Rs_i = (a_i*np.log((IL_i-i_mp)/Io_i+1) - v_mp)/i_mp
+    # params_i : initial values vector
+    params_i = np.array([IL_i, Io_i, a_i, Rsh_i, Rs_i])
+
+    # specs of module
+    specs = np.array([i_sc, v_oc, i_mp, v_mp, beta_voc, alpha_sc])
+
+    # computing
+    sol = root(pv_fct, x0=params_i, args=specs, method='lm').x
```

Comment:
Check the [`scipy.optimize.OptimizeResult`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.OptimizeResult.html#scipy.optimize.OptimizeResult) for success. I think the community decided to raise exceptions when fitting fails, e.g., [this example](https://github.com/pvlib/pvlib-python/blob/master/pvlib/ivtools.py#L107) and its corresponding tests.

Reply:
Good question. And these kind of parameter can easily result. For example if the user gives alpha_sc in A/K instead of %/K it will result in many cases in a negative resistance.

And I wonder if these negative resistances are really impossible ? I mean, as the single diode model is a model simplifying the complexity of the architecture of the semi-conductors layers in a PV cell, maybe a negative resistance could be the translation in the model of a physical interaction (generally minor in standard modules but more important in the module being calculated). Would it be possible that keeping this negative resistance would still improve the prediction made by the single diode model ? And finally, it would rather mean that SDM is not suitable for this particular module. 

But still, as what I say before must be rare, I think that a warning would be good to make the user check the units of the inputs.